File: .skills/openclaw-skills/skills/403914291/doc-publisher/src/wechat-api.js
/**
 * 微信公众号 API
 * 独立版本（不依赖 wechat-publisher）
 */

const https = require('https');
const path = require('path');
const fs = require('fs');

/**
 * 加载 .env 配置
 */
function loadEnv() {
  const envPath = path.join(__dirname, '..', '.env');
  if (fs.existsSync(envPath)) {
    const envContent = fs.readFileSync(envPath, 'utf8');
    envContent.split('\n').forEach(line => {
      const parts = line.split('=');
      if (parts[0] && parts[1]) process.env[parts[0].trim()] = parts[1].trim();
    });
  }
}
loadEnv();

/**
 * 微信配置
 */
const CONFIG = {
  APPID: process.env.WECHAT_APPID,
  SECRET: process.env.WECHAT_SECRET,
  THUMB_MEDIA_ID: process.env.WECHAT_THUMB_MEDIA_ID || ''
};

/**
 * 设置封面图片 ID
 */
function setThumbMediaId(mediaId) {
  CONFIG.THUMB_MEDIA_ID = mediaId;
}

/**
 * HTTPS 请求
 */
function request(options, body) {
  return new Promise((resolve, reject) => {
    const req = https.request(options, res => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        if (!data) return reject(new Error('Empty response'));
        try { resolve(JSON.parse(data)); }
        catch (e) { reject(new Error(e.message)); }
      });
    });
    req.on('error', reject);
    if (body) {
      const json = JSON.stringify(body);
      req.setHeader('Content-Type', 'application/json');
      req.setHeader('Content-Length', Buffer.byteLength(json));
      req.write(json);
    }
    req.end();
  });
}

/**
 * Token 缓存
 */
let tokenCache = { token: null, expires: 0 };

/**
 * 获取 Access Token
 */
async function getAccessToken() {
  const now = Date.now();
  if (tokenCache.token && now < tokenCache.expires) return tokenCache.token;
  
  const p = '/cgi-bin/token?grant_type=client_credential&appid=' + CONFIG.APPID + '&secret=' + CONFIG.SECRET;
  const res = await request({ hostname: 'api.weixin.qq.com', path: p, method: 'GET' });
  
  tokenCache = { token: res.access_token, expires: now + (res.expires_in - 300) * 1000 };
  return tokenCache.token;
}

/**
 * 创建草稿
 */
async function addDraft(article) {
  const token = await getAccessToken();
  
  const res = await request({
    hostname: 'api.weixin.qq.com',
    path: '/cgi-bin/draft/add?access_token=' + token,
    method: 'POST'
  }, {
    articles: [{
      title: article.title,
      author: article.author || 'Editor',
      digest: article.digest,
      content: article.content,
      thumb_media_id: CONFIG.THUMB_MEDIA_ID,
      show_cover_pic: 1
    }]
  });
  
  // 检查微信 API 错误
  if (res.errcode && res.errcode !== 0) {
    throw new Error(`微信 API 错误 ${res.errcode}: ${res.errmsg}`);
  }
  
  if (!res.media_id) {
    throw new Error('发布失败：未返回 media_id，响应：' + JSON.stringify(res));
  }
  
  return { media_id: res.media_id, created_at: Date.now(), title: article.title };
}

/**
 * 获取草稿列表
 */
async function getAllDrafts() {
  const token = await getAccessToken();
  
  const res = await request({
    hostname: 'api.weixin.qq.com',
    path: '/cgi-bin/draft/batchget?access_token=' + token,
    method: 'POST'
  }, {
    offset: 0,
    count: 50,
    no_content: 1
  });
  
  if (res.errcode && res.errcode !== 0) {
    throw new Error(`微信 API 错误 ${res.errcode}: ${res.errmsg}`);
  }
  
  return res.item || [];
}

/**
 * 删除草稿
 */
async function deleteDraft(mediaId) {
  const token = await getAccessToken();
  
  const res = await request({
    hostname: 'api.weixin.qq.com',
    path: '/cgi-bin/draft/delete?access_token=' + token,
    method: 'POST'
  }, { media_id: mediaId });
  
  if (res.errcode && res.errcode !== 0) {
    throw new Error(`微信 API 错误 ${res.errcode}: ${res.errmsg}`);
  }
  
  return true;
}

/**
 * 上传临时图片（用于封面等）
 */
async function uploadTempImage(imagePath) {
  const token = await getAccessToken();
  const fileContent = fs.readFileSync(imagePath);
  
  return new Promise((resolve, reject) => {
    const boundary = '----WebKitFormBoundary' + Math.random().toString(36).substring(2);
    const body = Buffer.concat([
      Buffer.from(`------WebKitFormBoundary\r\nContent-Disposition: form-data; name="media"; filename="cover.png"\r\nContent-Type: image/png\r\n\r\n`),
      fileContent,
      Buffer.from(`\r\n------WebKitFormBoundary\r\nContent-Disposition: form-data; name="description"\r\n\r\nLMStudio 封面\r\n------WebKitFormBoundary--\r\n`)
    ]);
    
    const req = https.request({
      hostname: 'api.weixin.qq.com',
      path: '/cgi-bin/material/add_material?access_token=' + token + '&type=image',
      method: 'POST',
      headers: {
        'Content-Type': 'multipart/form-data; boundary=' + boundary,
        'Content-Length': body.length
      }
    }, res => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const result = JSON.parse(data);
          console.log('上传结果:', result);
          if (result.errcode && result.errcode !== 0) {
            reject(new Error('上传失败：' + result.errmsg));
          } else {
            resolve(result.media_id);
          }
        } catch (e) {
          reject(new Error('解析失败：' + e.message));
        }
      });
    });
    
    req.on('error', reject);
    req.write(body);
    req.end();
  });
}

/**
 * 上传图片到微信素材库
 */
async function uploadImage(imagePath) {
  const token = await getAccessToken();
  
  return new Promise((resolve, reject) => {
    const boundary = '----WebKitFormBoundary' + Math.random().toString(36).substring(2);
    const formData = fs.readFileSync(imagePath);
    
    const body = [];
    body.push(Buffer.from('--' + boundary + '\r\n'));
    body.push(Buffer.from('Content-Disposition: form-data; name="media"; filename="' + path.basename(imagePath) + '"\r\n'));
    body.push(Buffer.from('Content-Type: image/jpeg\r\n'));
    body.push(Buffer.from('\r\n'));
    body.push(formData);
    body.push(Buffer.from('\r\n--' + boundary + '--\r\n'));
    
    const req = https.request({
      hostname: 'api.weixin.qq.com',
      path: '/cgi-bin/material/add_material?access_token=' + token + '&type=image',
      method: 'POST',
      headers: {
        'Content-Type': 'multipart/form-data; boundary=' + boundary
      }
    }, res => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          const result = JSON.parse(data);
          if (result.media_id) {
            resolve(result);
          } else {
            reject(new Error(result.errmsg || 'Upload failed'));
          }
        } catch (e) {
          reject(new Error(e.message));
        }
      });
    });
    
    req.on('error', reject);
    req.write(Buffer.concat(body));
    req.end();
  });
}

module.exports = {
  getAccessToken,
  addDraft,
  deleteDraft,
  getAllDrafts,
  uploadImage,
  uploadTempImage,
  request,
  CONFIG,
  setThumbMediaId
};

File: .skills/openclaw-skills/skills/zhuzhudewenjian/wechat-draft-publisher/scripts/wechat_api.py
"""
微信公众号 API 模块
支持：获取 access_token、上传图片素材、新增草稿、发布草稿
"""

import json
import os
import time
import requests


class WeChatAPI:
    """微信公众号 API 客户端"""

    BASE_URL = "https://api.weixin.qq.com/cgi-bin"

    def __init__(self, appid: str = None, appsecret: str = None):
        self.appid = appid or os.environ.get("WECHAT_APPID", "")
        self.appsecret = appsecret or os.environ.get("WECHAT_APPSECRET", "")
        self._access_token = None
        self._token_expires_at = 0

        if not self.appid or not self.appsecret:
            raise ValueError(
                "需要设置环境变量 WECHAT_APPID 和 WECHAT_APPSECRET，"
                "或在初始化时传入 appid 和 appsecret"
            )

    def get_access_token(self) -> str:
        """获取 access_token，自动缓存和刷新"""
        if self._access_token and time.time() < self._token_expires_at - 300:
            return self._access_token

        url = f"{self.BASE_URL}/token"
        params = {
            "grant_type": "client_credential",
            "appid": self.appid,
            "secret": self.appsecret,
        }
        resp = requests.get(url, params=params, timeout=10)
        data = resp.json()

        if "access_token" not in data:
            raise RuntimeError(f"获取 access_token 失败: {data}")

        self._access_token = data["access_token"]
        self._token_expires_at = time.time() + data.get("expires_in", 7200)
        return self._access_token

    def upload_article_image(self, image_path: str) -> str:
        """
        上传文章内容中的图片（返回 URL，用于文章正文内嵌图片）
        注意：此接口上传的图片不占用素材库额度
        """
        token = self.get_access_token()
        url = f"{self.BASE_URL}/media/uploadimg?access_token={token}"

        with open(image_path, "rb") as f:
            files = {"media": (os.path.basename(image_path), f, "image/png")}
            resp = requests.post(url, files=files, timeout=30)

        data = resp.json()
        if "url" not in data:
            raise RuntimeError(f"上传文章图片失败: {data}")

        return data["url"]

    def upload_thumb_image(self, image_path: str) -> str:
        """
        上传封面图片为永久素材（返回 media_id，用于草稿封面）
        """
        token = self.get_access_token()
        url = f"{self.BASE_URL}/material/add_material?access_token={token}&type=image"

        with open(image_path, "rb") as f:
            files = {"media": (os.path.basename(image_path), f, "image/png")}
            resp = requests.post(url, files=files, timeout=30)

        data = resp.json()
        if "media_id" not in data:
            raise RuntimeError(f"上传封面图片失败: {data}")

        return data["media_id"]

    def add_draft(
        self,
        title: str,
        content: str,
        thumb_media_id: str,
        author: str = "",
        digest: str = "",
        content_source_url: str = "",
    ) -> str:
        """
        新增草稿
        返回草稿的 media_id
        """
        token = self.get_access_token()
        url = f"{self.BASE_URL}/draft/add?access_token={token}"

        article = {
            "title": title,
            "author": author,
            "digest": digest,
            "content": content,
            "thumb_media_id": thumb_media_id,
            "need_open_comment": 1,
            "only_fans_can_comment": 0,
        }
        if content_source_url:
            article["content_source_url"] = content_source_url

        payload = {"articles": [article]}
        resp = requests.post(
            url,
            data=json.dumps(payload, ensure_ascii=False).encode("utf-8"),
            headers={"Content-Type": "application/json"},
            timeout=15,
        )

        data = resp.json()
        if "media_id" not in data:
            raise RuntimeError(f"新增草稿失败: {data}")

        return data["media_id"]

    def submit_publish(self, media_id: str) -> str:
        """
        发布草稿（提交发布任务）
        返回 publish_id
        """
        token = self.get_access_token()
        url = f"{self.BASE_URL}/freepublish/submit?access_token={token}"

        payload = {"media_id": media_id}
        resp = requests.post(url, json=payload, timeout=15)

        data = resp.json()
        if data.get("errcode", 0) != 0:
            raise RuntimeError(f"发布草稿失败: {data}")

        return data.get("publish_id", "")
